File: medicare_plus/utils/demo_data.py
# -*- coding: utf-8 -*-
"""
@license
SPDX-License-Identifier: Apache-2.0
"""
import json
import datetime
from database.db_manager import DatabaseManager

def seed_demo_data(db: DatabaseManager):
    # Check if settings are already populated
    existing_profile = db.fetch_one("SELECT * FROM health_profile LIMIT 1;")
    if existing_profile:
        print("[Demo Data] Database already seeded. Skipping initial loading.")
        return

    print("[Demo Data] Seeding SQLite database with high-quality patient logs.")

    # 1. Seed Health Profile
    db.execute_query("""
    INSERT INTO health_profile (name, age, gender, blood_type, height, weight, allergies, conditions, emergency_contact, emergency_phone)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?);
    """, (
        "Alex Mercer", 35, "Male", "O+", 180.0, 78.0,
        json.dumps(["Penicillin", "Peanuts"]),
        json.dumps(["Mild Hypertension"]),
        "Sarah Mercer", "+1 (555) 0199-382"
    ))

    # 2. Seed Medicines
    m1_id = db.execute_query("""
    INSERT INTO medicines (name, type, dosage, frequency, times, meal_relation, start_date, end_date, notes, color, refill_alert_days, is_active, created_at)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1, ?);
    """, ("Lisinopril", "Tablet", "10mg", "Daily", json.dumps(["08:00"]), "Before Meal", "2026-06-15", "2026-09-15", "For blood pressure control. Do not skip.", "#0D6E6E", 7, "2026-06-15T08:00:00"))

    m2_id = db.execute_query("""
    INSERT INTO medicines (name, type, dosage, frequency, times, meal_relation, start_date, end_date, notes, color, refill_alert_days, is_active, created_at)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1, ?);
    """, ("Metformin", "Capsule", "500mg", "Twice daily", json.dumps(["08:30", "20:30"]), "With Meal", "2026-06-15", "2026-12-15", "Take with breakfasts and dinners.", "#FF6B6B", 5, "2026-06-15T08:00:00"))

    m3_id = db.execute_query("""
    INSERT INTO medicines (name, type, dosage, frequency, times, meal_relation, start_date, end_date, notes, color, refill_alert_days, is_active, created_at)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1, ?);
    """, ("Atorvastatin", "Tablet", "20mg", "Daily", json.dumps(["21:00"]), "After Meal", "2026-06-16", "2026-12-16", "Cholesterol management. Take at night.", "#F39C12", 10, "2026-06-15T08:00:00"))

    # 3. Seed Compliance Logs (Past 5 days: June 17 - June 21)
    dates = ["2026-06-17", "2026-06-18", "2026-06-19", "2026-06-20", "2026-06-21"]
    
    for dt in dates:
        # Lisinopril
        status_lis = "skipped" if dt == "2026-06-17" else "missed" if dt == "2026-06-20" else "taken"
        taken_lis = "08:12" if status_lis == "taken" else ""
        db.execute_query("""
        INSERT INTO medicine_logs (medicine_id, scheduled_time, taken_time, status, date, notes)
        VALUES (?, ?, ?, ?, ?, ?);
        """, (m1_id, "08:00", taken_lis, status_lis, dt, "Doctor hold BP" if status_lis == "skipped" else ""))

        # Metformin morning
        status_met_m = "missed" if dt == "2026-06-18" else "taken"
        taken_met_m = "08:45" if status_met_m == "taken" else ""
        db.execute_query("""
        INSERT INTO medicine_logs (medicine_id, scheduled_time, taken_time, status, date, notes)
        VALUES (?, ?, ?, ?, ?, ?);
        """, (m2_id, "08:30", taken_met_m, status_met_m, dt, ""))

        # Metformin evening
        status_met_e = "missed" if dt == "2026-06-21" else "taken"
        taken_met_e = "20:35" if status_met_e == "taken" else ""
        db.execute_query("""
        INSERT INTO medicine_logs (medicine_id, scheduled_time, taken_time, status, date, notes)
        VALUES (?, ?, ?, ?, ?, ?);
        """, (m2_id, "20:30", taken_met_e, status_met_e, dt, ""))

        # Atorvastatin
        status_at = "missed" if dt == "2026-06-19" else "taken"
        taken_at = "21:10" if status_at == "taken" else ""
        db.execute_query("""
        INSERT INTO medicine_logs (medicine_id, scheduled_time, taken_time, status, date, notes)
        VALUES (?, ?, ?, ?, ?, ?);
        """, (m3_id, "21:00", taken_at, status_at, dt, ""))

    # 4. Seed Vitals (10 logs for BP and 10 logs for Blood Glucose)
    vitals_bp_data = [
        (135.0, 85.0, "2026-06-13 08:00", "2026-06-13", "Routine checkup"),
        (132.0, 84.0, "2026-06-14 08:15", "2026-06-14", ""),
        (128.0, 80.0, "2026-06-15 08:30", "2026-06-15", "Routine BP"),
        (134.0, 82.0, "2026-06-16 08:05", "2026-06-16", ""),
        (130.0, 81.0, "2026-06-17 08:10", "2026-06-17", "Routine BP check"),
        (126.0, 79.0, "2026-06-18 08:20", "2026-06-18", ""),
        (125.0, 78.0, "2026-06-19 08:00", "2026-06-19", ""),
        (138.0, 86.0, "2026-06-20 09:00", "2026-06-20", "Poor sleep high"),
        (129.0, 82.0, "2026-06-21 08:45", "2026-06-21", ""),
        (124.0, 80.0, "2026-06-22 08:00", "2026-06-22", "After workout")
    ]
    for sys, dia, rtime, dt, notes in vitals_bp_data:
        db.execute_query("""
        INSERT INTO vitals (type, value1, value2, unit, reading_time, notes, date)
        VALUES (?, ?, ?, ?, ?, ?, ?);
        """, ("bp", sys, dia, "mmHg", rtime, notes, dt))

    vitals_sugar_data = [
        (98.0, "2026-06-13 07:30", "2026-06-13", "Fasting"),
        (115.0, "2026-06-14 10:30", "2026-06-14", "Post breakfast"),
        (92.0, "2026-06-15 07:30", "2026-06-15", "Fasting"),
        (104.0, "2026-06-16 07:45", "2026-06-16", "Fasting"),
        (95.0, "2026-06-17 07:30", "2026-06-17", "Fasting"),
        (142.0, "2026-06-18 13:30", "2026-06-18", "After heavy lunch"),
        (89.0, "2026-06-19 07:35", "2026-06-19", "Fasting"),
        (91.0, "2026-06-20 07:40", "2026-06-20", "Fasting"),
        (106.0, "2026-06-21 07:50", "2026-06-21", "Fasting"),
        (96.0, "2026-06-22 07:30", "2026-06-22", "Fasting")
    ]
    for glu, rtime, dt, notes in vitals_sugar_data:
        db.execute_query("""
        INSERT INTO vitals (type, value1, value2, unit, reading_time, notes, date)
        VALUES (?, ?, ?, NULL, ?, ?, ?);
        """, ("sugar", glu, "mg/dL", rtime, notes, dt))

    # 5. Seed Appointments (2 Upcoming)
    db.execute_query("""
    INSERT INTO appointments (doctor_name, specialization, hospital, date, time, type, status, reminder_before, notes, prescription_path, created_at)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);
    """, ("Dr. Evelyn Thomas", "Cardiologist", "Metro Health Cardiology", "2026-06-24", "14:30", "in_person", "scheduled", 60, "Six-month follow up on blood pressure routine. Bring Lisinopril log.", "", "2026-06-22T08:00:00"))

    db.execute_query("""
    INSERT INTO appointments (doctor_name, specialization, hospital, date, time, type, status, reminder_before, notes, prescription_path, created_at)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);
    """, ("Dr. Aaron Patel", "Endocrinologist", "Valley Medical Specialities", "2026-07-10", "11:00", "video", "scheduled", 120, "Review HbA1c glucose levels. Virtual call via health portal.", "", "2026-06-22T08:00:00"))

    # 6. Seed Lab Report
    db.execute_query("""
    INSERT INTO lab_reports (title, date, doctor, file_path, notes, created_at)
    VALUES (?, ?, ?, ?, ?, ?);
    """, ("Lipid & CBC Panel Lab", "2026-06-10", "Dr. Evelyn Thomas", "report_lipid_june.pdf", "Cholesterol levels slightly elevated. HDL is 45, LDL is 135.", "2026-06-10T11:00:00"))

    # 7. Seed 3 Journal Entries
    db.execute_query("""
    INSERT INTO health_notes (date, mood, symptoms, note, created_at)
    VALUES (?, ?, ?, ?, ?);
    """, ("2026-06-19", 4, json.dumps(["Energy Level Normal"]), "Woke up refreshed. BP is stable. Had a 20 minute walk in the evening.", "2026-06-19T21:00:00"))

    db.execute_query("""
    INSERT INTO health_notes (date, mood, symptoms, note, created_at)
    VALUES (?, ?, ?, ?, ?);
    """, ("2026-06-20", 2, json.dumps(["Headache", "Mild Tension"]), "Had a stressful day at work. Slight headache, BP went up to 138/86. Drank chamomile tea.", "2026-06-20T21:00:00"))

    db.execute_query("""
    INSERT INTO health_notes (date, mood, symptoms, note, created_at)
    VALUES (?, ?, ?, ?, ?);
    """, ("2026-06-21", 5, json.dumps(["Energetic", "Restful Sleep"]), "Felt incredible today. Did some stretching and cooking.", "2026-06-21T21:30:00"))

    # 8. Add Settings Rows
    settings_rows = [
        ("patient_name", "Alex Mercer"),
        ("theme", "dark"),
        ("weight_unit", "kg"),
        ("temp_unit", "C"),
        ("time_format", "12h"),
        ("sound_enabled", "1"),
        ("notifications_enabled", "1"),
        ("quiet_hours_enabled", "0"),
        ("quiet_hours_start", "22:00"),
        ("quiet_hours_end", "07:00")
    ]
    for key, val in settings_rows:
        db.execute_query("INSERT OR REPLACE INTO settings (key, value) VALUES (?, ?);", (key, val))

    print("[Demo Data] Seeding successfully completed.")

File: src/utils/demoData.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

import { Medicine, MedicineLog, Appointment, Vital, HealthProfile, LabReport, HealthNote, Reminder } from '../types';

export function initializeDemoData() {
  if (localStorage.getItem('medicare_initialized')) {
    return; // Already initialized
  }

  // 1. Health Profile
  const defaultProfile: HealthProfile = {
    id: 1,
    name: 'Alex Mercer',
    age: 35,
    gender: 'Male',
    blood_type: 'O+',
    height: 180,
    weight: 78,
    allergies: ['Penicillin', 'Peanuts'],
    conditions: ['Mild Hypertension'],
    emergency_contact: 'Sarah Mercer',
    emergency_phone: '+1 (555) 0199-382'
  };
  localStorage.setItem('medicare_profile', JSON.stringify(defaultProfile));

  // 2. Medicines
  const defaultMedicines: Medicine[] = [
    {
      id: 1,
      name: 'Lisinopril',
      type: 'Tablet',
      dosage: '10mg',
      frequency: 'Daily',
      times: ['08:00'],
      meal_relation: 'Before Meal',
      start_date: '2026-06-15',
      end_date: '2026-09-15',
      notes: 'For blood pressure control. Do not skip.',
      color: '#0D6E6E', // Primary Teal
      refill_alert_days: 7,
      is_active: true,
      created_at: '2026-06-15T08:00:00Z'
    },
    {
      id: 2,
      name: 'Metformin',
      type: 'Capsule',
      dosage: '500mg',
      frequency: 'Twice daily',
      times: ['08:30', '20:30'],
      meal_relation: 'With Meal',
      start_date: '2026-06-15',
      end_date: '2026-12-15',
      notes: 'Take with breakfasts and dinners.',
      color: '#FF6B6B', // Accent Coral
      refill_alert_days: 5,
      is_active: true,
      created_at: '2026-06-15T08:00:00Z'
    },
    {
      id: 3,
      name: 'Atorvastatin',
      type: 'Tablet',
      dosage: '20mg',
      frequency: 'Daily',
      times: ['21:00'],
      meal_relation: 'After Meal',
      start_date: '2026-06-16',
      end_date: '2026-12-16',
      notes: 'Cholesterol management. Take at night.',
      color: '#F39C12', // Warning Orange
      refill_alert_days: 10,
      is_active: true,
      created_at: '2026-06-15T08:00:00Z'
    }
  ];
  localStorage.setItem('medicare_medicines', JSON.stringify(defaultMedicines));

  // 3. Medicine Logs (last 5 days)
  // Let's generate logs starting from 2026-06-17 to 2026-06-21. Today is 2026-06-22.
  const defaultLogs: MedicineLog[] = [];
  const dates = ['2026-06-17', '2026-06-18', '2026-06-19', '2026-06-20', '2026-06-21'];
  let logId = 1;

  dates.forEach((date) => {
    // Lisinopril (daily 08:00)
    // 17: skipped, 18: taken, 19: taken, 20: missed, 21: taken
    let status_lis: 'taken' | 'skipped' | 'missed' = 'taken';
    if (date === '2026-06-17') status_lis = 'skipped';
    if (date === '2026-06-20') status_lis = 'missed';

    defaultLogs.push({
      id: logId++,
      medicine_id: 1,
      scheduled_time: '08:00',
      taken_time: status_lis === 'taken' ? '08:12' : '',
      status: status_lis,
      date,
      notes: status_lis === 'skipped' ? 'Doctor advised to hold since BP was low.' : ''
    });

    // Metformin (twice daily 08:30, 20:30)
    // morning dose status
    let status_met_m: 'taken' | 'missed' = 'taken';
    if (date === '2026-06-18') status_met_m = 'missed';

    defaultLogs.push({
      id: logId++,
      medicine_id: 2,
      scheduled_time: '08:30',
      taken_time: status_met_m === 'taken' ? '08:45' : '',
      status: status_met_m,
      date,
      notes: ''
    });

    // evening dose status
    let status_met_e: 'taken' | 'missed' = 'taken';
    if (date === '2026-06-21') status_met_e = 'missed';

    defaultLogs.push({
      id: logId++,
      medicine_id: 2,
      scheduled_time: '20:30',
      taken_time: status_met_e === 'taken' ? '20:35' : '',
      status: status_met_e,
      date,
      notes: ''
    });

    // Atorvastatin (daily 21:00)
    // 17: taken, 18: taken, 19: missed, 20: taken, 21: taken
    let status_at: 'taken' | 'missed' = 'taken';
    if (date === '2026-06-19') status_at = 'missed';

    defaultLogs.push({
      id: logId++,
      medicine_id: 3,
      scheduled_time: '21:00',
      taken_time: status_at === 'taken' ? '21:10' : '',
      status: status_at,
      date,
      notes: ''
    });
  });

  localStorage.setItem('medicare_logs', JSON.stringify(defaultLogs));

  // 4. Vitals (10 Blood Pressure and 10 Sugar readings)
  const defaultVitals: Vital[] = [];
  let vitalId = 1;

  // Let's generate measurements from June 13 to June 22
  const bpReadings = [
    { sys: 135, dia: 85, hr: 72, time: '2026-06-13T08:00:00' },
    { sys: 132, dia: 84, hr: 76, time: '2026-06-14T08:15:00' },
    { sys: 128, dia: 80, hr: 70, time: '2026-06-15T08:30:00' },
    { sys: 134, dia: 82, hr: 75, time: '2026-06-16T08:05:00' },
    { sys: 130, dia: 81, hr: 73, time: '2026-06-17T08:10:00' },
    { sys: 126, dia: 79, hr: 71, time: '2026-06-18T08:20:00' },
    { sys: 125, dia: 78, hr: 68, time: '2026-06-19T08:00:00' },
    { sys: 138, dia: 86, hr: 80, time: '2026-06-20T09:00:00' },  // warning high day
    { sys: 129, dia: 82, hr: 74, time: '2026-06-21T08:45:00' },
    { sys: 124, dia: 80, hr: 72, time: '2026-06-22T08:00:00' }
  ];

  bpReadings.forEach((bp) => {
    defaultVitals.push({
      id: vitalId++,
      type: 'bp',
      value1: bp.sys,
      value2: bp.dia,
      unit: 'mmHg',
      reading_time: bp.time,
      notes: bp.sys > 135 ? 'A bit high, did not sleep well.' : 'Routine BP check',
      date: bp.time.split('T')[0]
    });
  });

  const sugarReadings = [
    { value: 98, time: '2026-06-13T07:30:00', notes: 'Fasting' },
    { value: 115, time: '2026-06-14T10:30:00', notes: 'Post breakfast' },
    { value: 92, time: '2026-06-15T07:30:00', notes: 'Fasting' },
    { value: 104, time: '2026-06-16T07:45:00', notes: 'Fasting' },
    { value: 95, time: '2026-06-17T07:30:00', notes: 'Fasting' },
    { value: 142, time: '2026-06-18T13:30:00', notes: 'After heavy lunch' },
    { value: 89, time: '2026-06-19T07:35:00', notes: 'Fasting' },
    { value: 91, time: '2026-06-20T07:40:00', notes: 'Fasting' },
    { value: 106, time: '2026-06-21T07:50:00', notes: 'Fasting' },
    { value: 96, time: '2026-06-22T07:30:00', notes: 'Fasting' }
  ];

  sugarReadings.forEach((sugar) => {
    defaultVitals.push({
      id: vitalId++,
      type: 'sugar',
      value1: sugar.value,
      unit: 'mg/dL',
      reading_time: sugar.time,
      notes: sugar.notes,
      date: sugar.time.split('T')[0]
    });
  });

  localStorage.setItem('medicare_vitals', JSON.stringify(defaultVitals));

  // 5. Appointments
  const defaultAppointments: Appointment[] = [
    {
      id: 1,
      doctor_name: 'Dr. Evelyn Thomas',
      specialization: 'Cardiologist',
      hospital: 'Metro Health Cardiology',
      date: '2026-06-24', // Day after tomorrow (June 24, 2026)
      time: '14:30',
      type: 'un_person' as any, // fallback in-person
      status: 'scheduled',
      reminder_before: 60,
      notes: 'Six-month follow up on blood pressure routine. Bring Lisinopril log.',
      prescription_path: '',
      created_at: '2026-06-22T08:00:00Z'
    },
    {
      id: 2,
      doctor_name: 'Dr. Aaron Patel',
      specialization: 'Endocrinologist',
      hospital: 'Valley Medical Specialities',
      date: '2026-07-10',
      time: '11:00',
      type: 'video',
      status: 'scheduled',
      reminder_before: 120,
      notes: 'Review HbA1c glucose levels. Virtual call via health portal.',
      prescription_path: '',
      created_at: '2026-06-22T08:00:00Z'
    }
  ];
  // Map type correctly
  defaultAppointments[0].type = 'in_person';
  localStorage.setItem('medicare_appointments', JSON.stringify(defaultAppointments));

  // 6. Lab Report
  const defaultReports: LabReport[] = [
    {
      id: 1,
      title: 'Lipid & CBC Panel Lab',
      date: '2026-06-10',
      doctor: 'Dr. Evelyn Thomas',
      file_path: 'report_lipid_june.pdf',
      notes: 'Cholesterol levels slightly elevated. HDL is 45, LDL is 135.',
      created_at: '2026-06-10T11:00:00Z'
    }
  ];
  localStorage.setItem('medicare_reports', JSON.stringify(defaultReports));

  // 7. Journal Entries (3 entries)
  const defaultJournal: HealthNote[] = [
    {
      id: 1,
      date: '2026-06-19',
      mood: 4,
      symptoms: ['Energy Level Normal'],
      note: 'Woke up refreshed. BP is stable. Had a 20 minute walk in the evening.',
      created_at: '2026-06-19T21:00:00Z'
    },
    {
      id: 2,
      date: '2026-06-20',
      mood: 2, // low / sad mood
      symptoms: ['Headache', 'Mild Tension'],
      note: 'Had a stressful day at work. Slight headache, BP went up to 138/86. Drank chamomile tea.',
      created_at: '2026-06-20T21:00:00Z'
    },
    {
      id: 3,
      date: '2026-06-21',
      mood: 5, // great
      symptoms: ['Energetic', 'Restful Sleep'],
      note: 'Felt incredible today. Did some stretching and cooking. Metformin evening dose was missed but feeling great.',
      created_at: '2026-06-21T21:30:00Z'
    }
  ];
  localStorage.setItem('medicare_journal', JSON.stringify(defaultJournal));

  // 8. Reminders
  const defaultReminders: Reminder[] = [
    {
      id: 1,
      type: 'medicine',
      reference_id: 1,
      message: 'Time to take Lisinopril 10mg (Before Meal)',
      scheduled_time: '2026-06-22T08:00:00',
      status: 'dismissed',
      created_at: '2026-06-22T08:00:00Z'
    },
    {
      id: 2,
      type: 'medicine',
      reference_id: 2,
      message: 'Time to take Metformin 500mg (With Meal)',
      scheduled_time: '2026-06-22T08:30:00',
      status: 'pending',
      created_at: '2026-06-22T08:30:00Z'
    }
  ];
  localStorage.setItem('medicare_reminders', JSON.stringify(defaultReminders));

  // Write initialization tag
  localStorage.setItem('medicare_initialized', 'true');
  localStorage.setItem('medicare_settings', JSON.stringify({
    theme: 'dark',
    weight_unit: 'kg',
    temp_unit: '°C',
    time_format: '12h',
    sound_enabled: true,
    notifications_enabled: true,
    quiet_hours_enabled: false
  }));
}
